orchestrator/utils/state.py

```diff
@@ -150,65 +150,85 @@ def _build_arguments(func: StepFunc | InputStepFunc, state: State) -> list:  # n
 
     Raises:
          KeyError: if requested argument is not in the state, or cannot be reconstructed as an initial domain model.
+         ValueError: if requested argument cannot be converted to the expected type.
 
     """
+
     sig = inspect.signature(func)
+    if not sig.parameters:
+        return []
+
+    def _convert_to_uuid(v: Any) -> UUID:
+        """Converts the value to a UUID instance if it is not already one."""
+        return v if isinstance(v, UUID) else UUID(v)
+
     arguments: list[Any] = []
-    if sig.parameters:
-        for name, param in sig.parameters.items():
-            # Ignore dynamic arguments. Mostly need to deal with `const`
-            if param.kind in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD):
-                logger.warning("*args and **kwargs are not supported as step params")
-                continue
-
-            # If we find an argument named "state" we use the whole state as argument to
-            # This is mainly to be backward compatible with code that needs the whole state...
-            # TODO: Remove this construction
-            if name == "state":
-                arguments.append(state)
-                continue
-
-            # Workaround for the fact that you can't call issubclass on typing types
+    for name, param in sig.parameters.items():
+        # Ignore dynamic arguments. Mostly need to deal with `const`
+        if param.kind in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD):
+            logger.warning("*args and **kwargs are not supported as step params")
+            continue
+
+        # If we find an argument named "state" we use the whole state as argument to
+        # This is mainly to be backward compatible with code that needs the whole state...
+        # TODO: Remove this construction
+        if name == "state":
+            arguments.append(state)
+            continue
+
+        # Workaround for the fact that you can't call issubclass on typing types
+        try:
+            is_subscription_model_type = issubclass(param.annotation, SubscriptionModel)
+        except Exception:
+            is_subscription_model_type = False
+
+        if is_subscription_model_type:
+            subscription_id = _get_sub_id(state.get(name))
+            if subscription_id:
+                sub_mod = param.annotation.from_subscription(subscription_id)
+                arguments.append(sub_mod)
```

Comment:
Not for this PR, but someday we should get rid of this endless sequence of appends and use 'yield' for cleaner code. (less noise imo).